Bluetooth: Controller: Consolidated set of pending changes

cvinayak opened the pull request:
Set of commits that has fixed assorted test failures in conformance testing.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Add defines for ext adv header field offsets

Add defines for Extended Advertising Common Extended Payload
Format field offsets.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix sync set_clear to return aux_ptr reference

Update ull_adv_sync_pdu_set_clear() to return aux_ptr
reference so that aux_ptr field for chain PDUs can be
pupulated.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Cleanup periodic adv chaining aux ptr population

Cleanup Periodic Advertising Chain PDUs auxiliary pointer
population, populate the back to back auxiliary pointer
field in thread context. Also, removed an unused code for
filling data.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Minimal time reservation for aux and sync

Use minimal time reservation for auxiliary and sync PDU
reception. A peer device could be scheduling multiple
advertising sets in a short duration with small PDUs hence
using the minimal time reservation would avoid skipping
closely scheduled reception of multiple auxiliary PDUs.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix skipped Periodic Sync radio events

Fix skipped Periodic Advertising Sync radio events due to
increased CPU use during pre-emption in LLL from a Scan
radio event to Periodic Advertising Sync radio event by
increasing EVENT_OVERHEAD_START_US value.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix missing AD Data Set fragment operation

Fix missing AD data set fragment operation implementation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix Periodic Adv Sync from preemption

Fix Periodic Advertising Synchronization from being
preempted by an overlapping Auxiliary PDU scan event.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Set/Clear function to enqueue aux PDU in caller

Refactor the ull_adv_aux_hdr_set_clear function so that the
caller of the function will enqueue the auxiliary PDU when
enqueuing the primary PDU.

This change is required to support chaining of AD Data where
parent auxiliary PDU will be added with aux ptr fields when
chain PDUs are appended.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Duplicate and add aux offset to chain PDU in ULL

Refactor PDU chaining implementation to add aux offset in
the aux pointer field in the parent PDU when duplicating
chain PDUs.

By adding the aux offsets in the ULL execution context the
LLL prepare execution in Extended and Periodic Advertising
does not need to have a O(n) while loop in LLL execution
context.

This will reduce CPU use in the highest priority ISR
compared to current implementation where chain PDUs where
duplicated in ULL execution context and aux offset populated
in LLL execution context.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Rename Kconfig to BT_CTLR_ADV_AUX_PDU_BACK2BACK

Rename Kconfig BT_CTLR_ADV_PDU_BACK2BACK to
BT_CTLR_ADV_AUX_PDU_BACK2BACK to represent use of
back-to-back for Extended Advertising PDU chaining support
similar to BT_CTLR_ADV_SYNC_PDU_BACK2BACK that is used to
represent the use of back-to-back for Periodic Advertising
PDU chaining support.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Add Extended Advertising Data fragment operation

Add implementation to support Extended Advertising AD Data
fragment operation of first, intermediate, last, unchanged
and complete AD data operation using the HCI LE Set Extended
Advertising Data Command.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Add Extended Scan Resp Data fragment operation

Add implementation to support Extended Scan Response Data
fragment operation of first, intermediate, last, unchanged
and complete data operation using the HCI LE Set Extended
Scan Response Data Command.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Minor refactor to move ULL after LL functions

Minor refactor to reorder and move ULL functions after the
LL functions.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Minor scan response implementation refactor

Minor refactor of scan response implementation to rename sr_pdu
to sr_pdu_prev, and aux_pdu to sec_pdu_prev etc.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Refactor DF impl to add aux offset in ULL context

Refactor DF implementation to add aux offset in the aux
pointer field of the parent PDU when chain PDU is added
and/or is duplicated when PDU contents are updated.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Minor refactor to move initialization

Minor refactor to move initialization close to locality of
reference.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix directed adv connection setup stall

When directed advertising and continuous scanning are
simultaneously active, and a connection request is received
by the directed advertising state then the connection was
not established due to the done event not being generated.

Done event from the directed advertising state was not
generated because the prepare pipeline only had resume
event which does not get scheduled when there is prepare
event of the continuous scan state.

The fix is to explicitly abort all pending prepare and
resume event in the pipeline for the directed advertising
state when connection request is received.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix continuous scan initiator from being aborted

Fix to prevent a transmitted CONNECT_REQ PDU from being
aborted mid-air when initiator is using continuous scan
window that do not use time reservation. A different state
or role that overlaps the connection initiation requesting
for an abort should be denied when CONNECT_REQ PDU has been
enqueued for tranmission by the radio.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -231,25 +263,176 @@ uint8_t ll_adv_sync_ad_data_set(uint8_t handle, uint8_t op, uint8_t len,
+#if defined(CONFIG_BT_CTLR_ADV_SYNC_PDU_LINK)
+		if (!err) {
+			/* Fragment into chain PDU if len > 191 bytes */
+			if (len > (PDU_AC_PAYLOAD_SIZE_MAX -
+				   PDU_AC_EXT_PAYLOAD_OVERHEAD)) {
```
This way to calculate max PDU size for a single extended advertising PDU is used at leas in two places `PDU_AC_PAYLOAD_SIZE_MAX - PDU_AC_EXT_PAYLOAD_OVERHEAD1`. Maybe use of a single named macro that covers the subtraction? I would improve readability.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -277,6 +460,168 @@ uint8_t ll_adv_sync_ad_data_set(uint8_t handle, uint8_t op, uint8_t len,
+		/* Flags */
+		*dptr_chain = 0U;
+
+		/* TODO: should CTEInfo be added to chain PDU due to AD data
```
CTE count is given by `HCI_LE_Set_Connectionless_CTE_Transmit_Parameters`, hence it is not altered due to change on PDUs count in periodic advertising chain.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -277,6 +460,168 @@ uint8_t ll_adv_sync_ad_data_set(uint8_t handle, uint8_t op, uint8_t len,
+		/* No TgtA flag */
+
+		/* CTEInfo flag */
+		if (hdr_chain->cte_info) {
```
How new PDU may have any of these flags already filled? It is allocated in line 490.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -277,6 +460,168 @@ uint8_t ll_adv_sync_ad_data_set(uint8_t handle, uint8_t op, uint8_t len,
+			(void)memcpy(dptr_chain, ad_overflow, ad_len_overflow);
+			dptr_chain += ad_len_overflow;
+
+			/* Reduce the AD data in the previous PDU */
```
"in the previous PDU"? Comment in lines 559-561 says about reduction of a AD data in current PDU.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -1069,18 +1429,28 @@ uint8_t ull_adv_sync_pdu_set_clear(struct lll_adv_sync *lll_sync,
+	/* ADI */
+	if (hdr_add_fields & ULL_ADV_PDU_HDR_FIELD_ADI) {
+		ter_hdr.adi = 1U;
+		/* return the size of ADI structure */
```
Missing information in a function description.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -1069,18 +1429,28 @@ uint8_t ull_adv_sync_pdu_set_clear(struct lll_adv_sync *lll_sync,
+		hdr_data = (uint8_t *)hdr_data + sizeof(uint8_t);
+		/* pick the reference to ADI param */
+		(void)memcpy(&adi, hdr_data, sizeof(struct pdu_adv_adi *));
+		/* return the pointer to ADI struct inside the PDU */
```
Missing information in a function description.

ppryga-nordic reviewed (changes requested):
I've done a very brief review of all commits except this one: b905598de4b1e0ec89278e78ff444b6775493f8f.

cvinayak committed:
Bluetooth: Controller: AD data fragment operation review rework

Added define PDU_AC_EXT_AD_DATA_LEN_MAX, maximum AD data
length possible in AUX_ADV_IND when all Common Extended
Header Format fields are populuted.

Updated function description to explain the added ADI and
Aux Ptr fields support in updating the auxiliary PDUs.

Remove the redundant CTE info population code unnecessary
for new chain PDU as a result of AD data overflow.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -1332,6 +1332,134 @@ void ull_adv_aux_done(struct node_rx_event_done *done)
+
+		/* Fill the aux offset in the parent PDU */
+		offs_us = PDU_AC_US(pdu->len, phy_s, phy_flags) + mafs_us;
+		ull_adv_aux_ptr_fill(aux_ptr, offs_us, phy_s);
```
added function description
